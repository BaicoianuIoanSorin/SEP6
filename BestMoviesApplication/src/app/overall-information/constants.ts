export enum SearchByEnum {
  YEAR,
  RATING,
}

export enum LoadEnum {
  BEST_MOVIES,
  MOVIES_BY_YEAR,
  MOVIES_BY_RATING,
}
