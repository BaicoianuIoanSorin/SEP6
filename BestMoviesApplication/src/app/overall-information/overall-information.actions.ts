export class OverAllInformationFetchInfo {
  static readonly type = 'Overall Information - Fetching info';
  constructor() {
  }
}

export class OverAllInformationReset {
  static readonly type = 'Overall Information - Reset';
  constructor() {
  }
}
