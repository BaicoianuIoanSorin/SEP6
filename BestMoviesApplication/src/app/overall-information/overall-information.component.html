<ng-container *ngIf="isFetching$ | async; then dataFetching; else dataFetched"></ng-container>

<ng-template #dataFetched>
  <div class="overall-information"
       *ngIf="{movies: movies$ | async,
               people: people$ | async,
               bestMovies: bestMovies$ | async} as info">
    <div class="best-movies">
      <nb-card>
        <nb-card-header>
          Best movies
        </nb-card-header>
        <nb-card-body>
          <nb-list>
            <nb-list-item
              *ngFor="let movie of info.bestMovies">
              {{movie.year}}
            </nb-list-item>
          </nb-list>
        </nb-card-body>
      </nb-card>
    </div>
    <div class="general-information">
      <nb-card>
        <nb-card-header>
          Overall information
        </nb-card-header>
        <nb-card-body>
          <div>
            <nb-icon
              icon="film-outline">
            </nb-icon>
            Movies: {{info.movies.length}}
          </div>
          <div>
            <nb-icon
              icon="people-outline">
            </nb-icon>
            People: {{info.people.length}}
          </div>
        </nb-card-body>
      </nb-card>
    </div>
  </div>
</ng-template>

<ng-template #dataFetching>
  <nb-spinner
    [nbSpinner]="true"
    status="basic"
    class="spinner">
  </nb-spinner>
</ng-template>
