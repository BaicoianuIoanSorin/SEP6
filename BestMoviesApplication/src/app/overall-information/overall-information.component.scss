nb-card-header {
  text-align: center;
}

.overall-information {
  flex-direction: row;
  display: flex;
  grid-gap: 1rem;

  .best-movies {
    width:100%;
  }

  .general-information {
    width: 37%;

    nb-card-body {
      display: grid;
      gap: 0.4rem;
    }
  }
}

.spinner {
  color: nb-theme(color-primary-500);
}
