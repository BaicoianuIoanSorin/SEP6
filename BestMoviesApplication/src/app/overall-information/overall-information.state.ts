import {Movie} from "../../model/movie";
import {Person} from "../../model/person";
import {Action, State, StateContext} from "@ngxs/store";
import {Injectable} from "@angular/core";
import {
  OverAllInformationAddMovieToFavourites,
  OverAllInformationFetchBestMoviesTop,
  OverAllInformationFetchMoviesFromSameYear,
  OverAllInformationFetchSameRatingRange,
  OverAllInformationReset
} from "./overall-information.actions";
import {produce} from "immer";
import {NbToastrService} from "@nebular/theme";
import { MoviesService } from "src/api/movies.service";
import { throwError } from "rxjs";
import { catchError, tap } from "rxjs/operators";
import {FavoritesService} from "src/api/favorites.service";
import {HttpErrorResponse} from "@angular/common/http";

export interface OverAllInformationStateModel {
  isFetching: boolean;
  moviesByYearIsFiltered: boolean;
  movies: Movie[];
  people: Person[];
  topMovies: Movie[];
  topMoviesByRating: Movie[];
  topMoviesByYear: Movie[];
}

export const defaultsState: OverAllInformationStateModel = {
  isFetching: false,
  moviesByYearIsFiltered: false,
  movies: [],
  people: [],
  topMovies: [],
  topMoviesByRating: [],
  topMoviesByYear: [],
}

@State<OverAllInformationStateModel>( {
  name: 'overallInformationPage',
  defaults: defaultsState,
})

@Injectable()
export class OverAllInformationState {
  movies: Movie[] = [];
  people: Person[] = [];

  constructor(
    private nbToastrService: NbToastrService,
    private moviesService: MoviesService,
    private favoritesService: FavoritesService,
  ) {
  }

  @Action(OverAllInformationFetchBestMoviesTop)
  overAllInformationFetchBestMoviesTop(
    {getState, setState}: StateContext<OverAllInformationStateModel>,
    action: OverAllInformationFetchBestMoviesTop) {

    let newState = produce(getState(), draft => {
      draft.isFetching = true;
    })
    setState(newState);

    this.moviesService.getNMostPopularMovies(action.top)
    .pipe(
      tap((movies: Movie[]) => {
        newState = produce(getState(), draft => {
          draft.topMovies = movies;
          draft.isFetching = false;
        });
        setState(newState);
      }),
      catchError((error) => {
        newState = produce(getState(), draft => {
          draft.isFetching = false;
        });
        setState(newState);
        return throwError(error);
      })
    )
    .subscribe();
  }

  @Action(OverAllInformationAddMovieToFavourites)
  overAllInformationAddMovieToFavourites(
    {getState, setState}: StateContext<OverAllInformationStateModel>,
    action: OverAllInformationAddMovieToFavourites) {

    let newState = produce(getState(), draft => {
      draft.isFetching = true;
    })
    setState(newState);

    this.favoritesService.addMoviesToFavorites(action.userName, action.movieId)
      .pipe(
        tap(() => {
          newState = produce(getState(), draft => {
            draft.isFetching = false;
          });
          setState(newState);
        }),
        catchError((error: HttpErrorResponse) => {
          if(error.status === 400) {
            this.nbToastrService.show('Movie already existing in favourite list', 'You cannot add this movie in favourite list', { status: 'warning'})
          }
          newState = produce(getState(), draft => {
            draft.isFetching = false;
          });
          setState(newState);
          return throwError(error);
        })
      )
      .subscribe();
  }


  @Action(OverAllInformationFetchSameRatingRange)
  overAllInformationFetchSameRatingRange(
    {getState, setState}: StateContext<OverAllInformationStateModel>,
    action: OverAllInformationFetchSameRatingRange) {

    let newState = produce(getState(), draft => {
      draft.isFetching = true;
    })
    setState(newState);

    if (!action.rating) {
      newState = produce(getState(), draft => {
        draft.isFetching = false;
      })

      setState(newState);
      return this.nbToastrService.show(
        "Rating is not found",
        "Couldn't fetch information about movies from same rating",
        {
          status: "warning"
        }
      );
    }

    this.moviesService.getNMoviesByRating(action.rating, action.listSize)
    .pipe(
      tap((movies: Movie[]) => {
        newState = produce(getState(), draft => {
          draft.topMoviesByRating = movies;
          draft.isFetching = false;
        })
        setState(newState);
      }),
      catchError((error) => {
        newState = produce(getState(), draft => {
          draft.isFetching = false;
        });
        setState(newState);
        return throwError(error);
      })
    )
    .subscribe();
  }


  @Action(OverAllInformationFetchMoviesFromSameYear)
  overAllInformationFetchMoviesFromSameYear(
    {getState, setState}: StateContext<OverAllInformationStateModel>,
    action: OverAllInformationFetchMoviesFromSameYear) {

    let newState = produce(getState(), draft => {
      draft.isFetching = true;
    })
    setState(newState);

    if (!action.year) {
      newState = produce(getState(), draft => {
        draft.isFetching = false;
      })
      setState(newState);
      return this.nbToastrService.show(
        "Movie is not found",
        "Couldn't fetch information about movies from same year",
        {
          status: "warning"
        }
      );
    }

    this.moviesService.getNMoviesByYear(action.year, action.listSize)
    .pipe(
      tap((movies: Movie[]) => {
        newState = produce(getState(), draft => {
          draft.topMoviesByYear = movies;
          draft.isFetching = false;
        })
        setState(newState);
      }),
      catchError((error) => {
        newState = produce(getState(), draft => {
          draft.isFetching = false;
        });
        setState(newState);
        return throwError(error);
      })
    )
    .subscribe();
  }


  @Action(OverAllInformationReset)
  async overAllInformationReset(
    { getState, setState }: StateContext<OverAllInformationStateModel>,
    action: OverAllInformationReset) {
    setState(defaultsState);
  }

}
