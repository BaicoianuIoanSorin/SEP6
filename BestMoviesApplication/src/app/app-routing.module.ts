import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import {OverallInformationComponent} from "./overall-information/overall-information.component";
import {NotFoundComponent} from "src/app/not-found/not-found.component";
import {MoviesComponent} from "src/app/information/movies/movies.component";
import {PeopleComponent} from "src/app/information/people/people.component";

const routes: Routes = [
  {
    path: 'overall-information',
    title: 'Overall Information',
    component: OverallInformationComponent,
  },
  {
    path: 'information',
    children: [
      {
        path:'movies',
        title: 'Information Movies',
        component: MoviesComponent,
      },
      {
        path:'people',
        title: 'Information People',
        component: PeopleComponent,
      }
    ]
  },
  { path: '',
    redirectTo: '/overall-information',
    pathMatch: 'full'
  },
  {
    path: '**',
    title: '404 Page not found',
    component: NotFoundComponent
  }
];

@NgModule({
  imports: [RouterModule.forRoot(routes)],
  exports: [RouterModule]
})
export class AppRoutingModule { }
