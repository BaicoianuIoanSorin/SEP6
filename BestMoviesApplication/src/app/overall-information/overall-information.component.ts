import {Component, OnDestroy, OnInit, ViewChild} from '@angular/core';
import {Select, Selector, Store} from "@ngxs/store";
import {OverallInformationSelector} from "./overall-information.selector";
import {Observable} from "rxjs";
import {Movie} from "../../model/movie";
import {Person} from "../../model/person";
import {
  OverAllInformationBestMoviesFetch,
  OverAllInformationFetchInfo,
  OverAllInformationReset
} from "./overall-information.actions";
import {CellClickedEvent, ColDef, ColumnApi, GridApi, GridReadyEvent} from "ag-grid-community";
import {AgGridAngular} from "ag-grid-angular";
import {Router} from "@angular/router";
import {
  MovieOverviewFetchBestMoviesTop,
  MovieOverviewFetchMoviesFromSameYear
} from "src/app/information/movies/movie-overview/movies-overview.actions";
import {MoviesOverviewSelector} from "src/app/information/movies/movie-overview/movies-overview.selector";
import {NbToastrService} from "@nebular/theme";

@Component({
  selector: 'app-overall-information',
  templateUrl: './overall-information.component.html',
  styleUrls: ['./overall-information.component.scss']
})
export class OverallInformationComponent implements OnInit, OnDestroy {

  alive = true;

  @Select(OverallInformationSelector.isFetching)
  isFetching$: Observable<boolean>;

  @Select(OverallInformationSelector.movies)
  movies$: Observable<Movie[]>;

  @Select(OverallInformationSelector.people)
  people$: Observable<Person[]>;

  @Select(MoviesOverviewSelector.isFetching)
  isFetchingBestMovies$: Observable<boolean>;

  @Select(MoviesOverviewSelector.topMovies)
  bestMovies$: Observable<Movie[]>;

  @Select(MoviesOverviewSelector.topMoviesByYear)
  bestMoviesByYear$: Observable<Movie[]>;

  gridApi: GridApi;
  gridColumnApi: ColumnApi;
  @ViewChild('agGridAngular') agGrid!: AgGridAngular;

  TOP_BEST_MOVIES: string = '5';
  TOP_BEST_MOVIES_YEAR: string = '5';
  YEAR_NOW: number;
  YEAR_SEARCHED: number;
  moviesByYearIsFiltered: boolean = false;
  placeHolder: string;


  constructor(private store: Store,
              private router: Router,
              private nbToastrService: NbToastrService) {
  }

  async ngOnInit() {
    let date = new Date();
    this.YEAR_NOW = date.getUTCFullYear();
    this.YEAR_SEARCHED = this.YEAR_NOW;
    this.placeHolder = `Year searched ${this.YEAR_SEARCHED}`;
    const actionsInParallel = [
      new OverAllInformationFetchInfo(),
      new OverAllInformationBestMoviesFetch(),
      new MovieOverviewFetchBestMoviesTop(Number(this.TOP_BEST_MOVIES)),
      new MovieOverviewFetchMoviesFromSameYear(Number(this.TOP_BEST_MOVIES_YEAR), this.YEAR_NOW),
    ];
    this.store.dispatch([...actionsInParallel]);
  }

  onSearchMoviesByYear(event) {
    this.YEAR_SEARCHED = Number(event);
    if (isNaN(this.YEAR_SEARCHED)) {
      return this.nbToastrService.show(`Searched for ${event}'`, `You only need to include numbers in search`, {
        status: 'warning'
      })
    }
    else {
      this.moviesByYearIsFiltered = true;
      this.placeHolder = `Year searched ${this.YEAR_SEARCHED}`;
      this.store.dispatch(new MovieOverviewFetchMoviesFromSameYear(Number(this.TOP_BEST_MOVIES_YEAR), this.YEAR_SEARCHED));
    }
  }

  resetSearchMoviesByYear() {
    this.moviesByYearIsFiltered = false;
    this.YEAR_SEARCHED = this.YEAR_NOW;
    this.placeHolder = `Year searched ${this.YEAR_SEARCHED}`;
    this.store.dispatch(new MovieOverviewFetchMoviesFromSameYear(Number(this.TOP_BEST_MOVIES_YEAR), this.YEAR_NOW));
  }

  redirectToMovieOverviewPage(movieId: number) {
    this.router.navigate([`/information/movies/${movieId}`]);
  }

  loadBestMovies() {
    this.store.dispatch(new MovieOverviewFetchBestMoviesTop(Number(this.TOP_BEST_MOVIES)));
  }

  loadBestMoviesByYear() {
    this.store.dispatch(new MovieOverviewFetchMoviesFromSameYear(Number(this.TOP_BEST_MOVIES_YEAR), this.YEAR_NOW));
  }

  ngOnDestroy() {
    this.alive = false;
    this.store.dispatch(new OverAllInformationReset());
  }
}
