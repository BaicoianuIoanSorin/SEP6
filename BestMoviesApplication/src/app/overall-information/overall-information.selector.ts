import {Selector} from "@ngxs/store";
import {OverAllInformationState, OverAllInformationStateModel} from "./overall-information.state";

export class OverallInformationSelector {
  @Selector([OverAllInformationState])
  static isFetching(state: OverAllInformationStateModel) {
    return state.isFetching;
  }

  @Selector([OverAllInformationState])
  static moviesByYearIsFiltered(state: OverAllInformationStateModel) {
    return state.moviesByYearIsFiltered;
  }

  @Selector([OverAllInformationState])
  static movies(state: OverAllInformationStateModel) {
    return state.movies;
  }

  @Selector([OverAllInformationState])
  static people(state: OverAllInformationStateModel) {
    return state.people;
  }
}
