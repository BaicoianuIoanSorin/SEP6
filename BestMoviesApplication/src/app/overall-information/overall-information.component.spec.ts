import { OverallInformationComponent } from './overall-information.component';
import {BmLayoutComponent} from "../../core/components/bm-layout/bm-layout.component";
import {Spectator} from "@ngneat/spectator";
import {createComponentFactory, SpectatorFactory} from "@ngneat/spectator/jest";
import {RouterTestingModule} from "@angular/router/testing";
import {BrowserModule} from "@angular/platform-browser";
import {AppRoutingModule} from "../app-routing.module";
import {BrowserAnimationsModule} from "@angular/platform-browser/animations";
import {
  NbSidebarService,
} from "@nebular/theme";
import {NEBULAR_MODULES, PROVIDERS, STATES} from "../app.module";
import {HttpClientModule} from "@angular/common/http";
import {NgxsModule} from "@ngxs/store";
import {environment} from "../../environments/environment";

describe('app > overall-information > overall-information.component.spec.ts', () => {
  let component: OverallInformationComponent;
  let spectator: Spectator<OverallInformationComponent>;
  let createComponent: SpectatorFactory<OverallInformationComponent> = createComponentFactory({
    component: OverallInformationComponent,
    imports: [
      RouterTestingModule,
      BrowserModule,
      HttpClientModule,
      AppRoutingModule,
      BrowserAnimationsModule,
      NgxsModule.forRoot(STATES, {
        developmentMode: !environment.production,
      }),
      ...NEBULAR_MODULES,
    ],
    declarations: [BmLayoutComponent],
    providers: PROVIDERS,
  });

  beforeEach(() => {
    spectator = createComponent();
  });

  it('should create', () => {
    expect(spectator.component).toBeTruthy();
  });

  it('should match snapshot', async () => {
    spectator.detectChanges();
    await spectator.fixture.whenStable();
    spectator.detectChanges();
    await spectator.fixture.whenRenderingDone();

    // assert
    expect(spectator.fixture).toMatchSnapshot();
  });
});
