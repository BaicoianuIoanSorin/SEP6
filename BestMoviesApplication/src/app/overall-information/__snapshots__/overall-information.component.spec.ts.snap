// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app > overall-information > overall-information.component.spec.ts should match snapshot 1`] = `
<app-overall-information
  TOP_BEST_MOVIES={[Function String]}
  TOP_BEST_MOVIES_YEAR={[Function String]}
  YEAR={[Function Number]}
  __isFetching$__selector={[Function Observable]}
  __isFetchingBestMovies$__selector={[Function Observable]}
  agGrid="undefined"
  alive={[Function Boolean]}
  router={[Function Router]}
  store={[Function Store]}
>
  <nb-spinner
    class="spinner size-medium status-basic nb-spinner-container"
    status="basic"
  >
    <span
      class="spin-circle"
    />
    <span
      class="message"
    >
      Loading...
    </span>
    <nb-spinner
      class="size-medium status-basic"
    >
      <span
        class="spin-circle"
      />
    </nb-spinner>
  </nb-spinner>
</app-overall-information>
`;
